### PR TITLE
Implement JsonCursor using the new Cursor API

### PR DESCRIPTION
**Pull request by jenn-le**
- Adds methods to `CursorAdapter` to allow `StackCursors` to be templated over any type instead of just `NodeData`
- Implement the previous `JsonCursor` as a factory function that creates a `StackCursor` using a custom adapter
- Moves old `JsonCursor` code and its tests into legacy files

Note: `jsonCursorLegacy.ts` is unchanged from the previous `jsonCursor.ts` (same with the corresponding test files)

### CODE REVIEW

File: packages/dds/tree/src/feature-libraries/treeCursorUtils.ts [hunk 1]
```diff
@@ -4,37 +4,34 @@
...
 ): ITreeCursorSynchronous {
     return new StackCursor(root, adapter);
 }
 
 /**
- * Provides functionality to allow `singleStackTreeCursor` to implement a cursor.
+ * Provides functionality to allow a {@link singleStackTreeCursor} to implement a cursor.
  */
 export interface CursorAdapter<TNode> {
+    value(node: TNode): any;
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/treeCursorUtils.ts:24]: Do not use `any`. `Value` should be what you want here.

File: packages/dds/tree/src/feature-libraries/treeCursorUtils.ts [hunk 2]
```diff
@@ -250,12 +244,19 @@ class StackCursor<TNode extends NodeData>
         return field;
     }
 
-    public get value(): Value {
-        return this.getNode().value;
+    /**
+     * @returns the value of the current node
+     */
+    public get value(): any {
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/treeCursorUtils.ts:250]: use `Value` not `any`

File: packages/dds/tree/src/feature-libraries/treeCursorUtils.ts [hunk 3]
```diff
@@ -250,12 +244,19 @@ class StackCursor<TNode extends NodeData>
...
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/treeCursorUtils.ts:251]: linter is right. Returning `any` is bad!
```suggestion
```

**Comment by CraigMacomber** [comment]
How does this extra indirection impact performance? Specifically whats the before and after for the "sum" benchmarks (they request values on all the nodes).

File: packages/dds/tree/src/domains/json/jsonCursorLegacy.ts [hunk 4]
```diff
@@ -0,0 +1,227 @@
+/*!
```
Comment (CraigMacomber) [inline packages/dds/tree/src/domains/json/jsonCursorLegacy.ts:1]: I'd prefer if this file move was not in the same PR as the implementation logic so the diff was better.

Anyway, is it safe to assume this file was just moved, and not edited?

File: packages/dds/tree/src/domains/json/jsonCursor.ts [hunk 5]
```diff
@@ -204,19 +96,19 @@ export function cursorToJsonObject(reader: ITreeCursor): JsonCompatible {
         case jsonString.name:
             return reader.value as number | boolean | string;
```
Comment (CraigMacomber) [inline packages/dds/tree/src/domains/json/jsonCursor.ts:97]:
```suggestion
        case jsonString.name: {
            assert(isPrimitiveValue(reader.value);
            return reader.value;
        }
```

File: packages/dds/tree/src/test/domains/json/jsonCursor.spec.ts [hunk 6]
```diff
@@ -64,35 +61,39 @@ describe("JsonCursor", () => {
...
+            assert.deepEqual(getKeysAsSet(singleJsonCursor({ x: {} })), new Set(["x"]));
             assert.deepEqual(
-                new Set(new JsonCursor({ x: {}, test: 6 }).keys),
+                getKeysAsSet(singleJsonCursor({ x: {}, test: 6 })),
                 new Set(["x", "test"]),
             );
         });
 
         it("array", () => {
             // TODO: should empty arrays report this key?
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/domains/json/jsonCursor.spec.ts:78]: The cursor API currently doesn't clarify this. We should fix that. I think explicitly documenting that only non-empty fields will be included in the iteration is reasonable. The other implementations attempt to ensure that, and your own cursorToJsonObject asserts if it finds an empty field on an object, so we should probably require that in the cursor API, fix empty arrays here, and update the tests accordingly.

If you want to update the existing generic cursor tests and API docs in a separate PR and land that before this one (and only do the new cursor API), that might be best.

**Comment by jenn-le** [comment]
> How does this extra indirection impact performance? Specifically whats the before and after for the "sum" benchmarks (they request values on all the nodes).

Ah right, I forgot to mention that. The benchmarks are all generally about the same with a few being a little faster using StackCursor, including the "sum" benchmarks for the few runs I tried.

File: packages/dds/tree/src/domains/json/jsonCursorLegacy.ts (hunk 4 above)
Comment (jenn-le) [reply packages/dds/tree/src/domains/json/jsonCursorLegacy.ts:1]: Yup, I made a note about it in the description. I can rename the new one if you think that would help, assuming that Github would detect the file rename as long as its old name isn't still being used.

Comment (CraigMacomber) [reply packages/dds/tree/src/domains/json/jsonCursorLegacy.ts:1]: You can leave it as is. Delta between the old and new implementations ended up being useful anyway.

File: packages/dds/tree/src/test/domains/json/jsonCursor.spec.ts (hunk 6 above)
Comment (jenn-le) [reply packages/dds/tree/src/test/domains/json/jsonCursor.spec.ts:78]: Addressed in #12375

File: packages/dds/tree/src/domains/json/jsonCursor.ts [hunk 7]
```diff
@@ -149,55 +49,42 @@ export class JsonCursor<T> implements ITreeCursor<SynchronousNavigationResult> {
...
+    keysFromNode: (node: JsonCompatible): readonly FieldKey[] => {
+        switch (typeof node) {
             case "object":
                 if (node === null) {
                     return [];
                 } else if (Array.isArray(node)) {
                     return [EmptyKey];
                 } else {
-                    return Object.keys(node as object) as Iterable<FieldKey>;
+                    return Object.getOwnPropertyNames(node as object) as FieldKey[];
```
Comment (CraigMacomber) [inline packages/dds/tree/src/domains/json/jsonCursor.ts:61]: Is this inner type conversion needed? Typescript MIGHT be able to infer this from the switch statement;
```suggestion
                    return Object.getOwnPropertyNames(node) as FieldKey[];
```

Also, Object.keys is fine and probably better here. It returns a subset of what getOwnPropertyNames does, since it skips non enumerable properties. (object.keys does NOT return non own properties, the names are misleading: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys )

We don't expect our object to have non enumerable properties (like the length of an array), but if it did, they should probably be skipped (or error). 

```suggestion
                    return Object.keys(node) as FieldKey[];
```

File: packages/dds/tree/src/domains/json/jsonCursor.ts [hunk 8]
```diff
@@ -149,55 +49,42 @@ export class JsonCursor<T> implements ITreeCursor<SynchronousNavigationResult> {
...
-        return (node as JsonCompatibleObject)[key as LocalFieldKey] === undefined
-            ? 0 // A field with an undefined value has 0 length
-            : 1; // All other fields have a length of 1
-    }
-
-    public currentFieldLength(): number {
-        const parent = this.parentStack[this.parentStack.length - 1];
-        if (this.currentKey === EmptyKey && Array.isArray(parent)) {
-            return parent.length;
+        if (Object.prototype.hasOwnProperty.call(node, key)) {
```
Comment (CraigMacomber) [inline packages/dds/tree/src/domains/json/jsonCursor.ts:72]: I think (add a test for it), this mishandles requesting a property named "0" on an array thats not empty.

We should instead do someting like:
```
 if (Array.isArray(node)) {
            return key === EmptyKey ?? node [];
  }
```

to prevent arrays from falling through here

File: packages/dds/tree/src/domains/json/jsonCursor.ts [hunk 9]
```diff
@@ -149,55 +49,42 @@ export class JsonCursor<T> implements ITreeCursor<SynchronousNavigationResult> {
...
+            const field = (node as JsonCompatibleObject)[key as LocalFieldKey];
+            return field === undefined ? [] : [field];
```
Comment (CraigMacomber) [inline packages/dds/tree/src/domains/json/jsonCursor.ts:74]: Since the node object is supposed to be json compatible, and explicit undefined fields are not preserved in json (and make keysFromNode incorrect due to empty fields), I think we should assert not undefined here instead of handling it.

File: packages/dds/tree/src/feature-libraries/treeCursorUtils.ts [hunk 10]
```diff
@@ -177,6 +178,12 @@ class StackCursor<TNode extends NodeData>
             this.exitField();
             return false;
         }
+
+        // Skip empty fields during iteration
+        if (this.getField().length === 0) {
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/treeCursorUtils.ts:183]: Have you benchmarked how much overhead this (and the one below) adds for MapTree and text Cursors? They currently avoid empty keys so they shouldn't need this case: if it slows them down a few percent we might want to remove and require keysFromNode only include empty fields.

**Review by CraigMacomber — APPROVED** [review]

File: packages/dds/tree/src/feature-libraries/treeCursorUtils.ts (hunk 10 above)
Comment (jenn-le) [reply packages/dds/tree/src/feature-libraries/treeCursorUtils.ts:183]: Seems like it slows down some of the benchmarks significantly. I think putting the burden on keysFromNode makes sense. I'll do that and benchmark the difference.

Before:

```
✔   Direct: 'canada'                                  5,269,672.7                    ±1.50%         900       75            5.45
✔   cursorToJsonObject(JsonCursor): 'canada'         26,397,734.9                    ±5.77%         153       51            5.68
✔   cursorToJsonObject(TextCursor): 'canada'         32,707,442.5                    ±3.86%         108       54            5.59
✔   cursorToJsonObject(MapCursor): 'canada'          31,026,583.5                    ±3.85%         114       57            5.64
✔   jsonableTreeFromCursor(JsonCursor): 'canada'     35,852,681.0                    ±3.81%          96       48            5.61
✔   jsonableTreeFromCursor(TextCursor): 'canada'     44,378,280.8                    ±4.24%          82       41            5.74
✔   jsonableTreeFromCursor(MapCursor): 'canada'      49,556,620.8                    ±6.72%          76       38            5.72
✔   mapTreeFromCursor(JsonCursor): 'canada'          59,641,893.9                   ±10.36%          66       33            5.79
✔   mapTreeFromCursor(TextCursor): 'canada'          59,500,950.2                    ±6.35%          72       36            5.63
✔   mapTreeFromCursor(MapCursor): 'canada'           61,002,783.3                    ±6.84%          80       40            5.68
✔   sum(JsonCursor): 'canada'                         9,392,415.7                    ±0.41%         483       69            5.45
✔   sum(TextCursor): 'canada'                        13,200,300.0                    ±0.50%         330       66            5.51
✔   sum(MapCursor): 'canada'                         15,164,852.1                    ±0.45%         272       68            5.45
✔   sum-map(JsonCursor): 'canada'                    17,616,136.8                    ±0.40%         219       73            5.45
✔   sum-map(TextCursor): 'canada'                    20,780,093.2                    ±0.34%         189       63            5.53
✔   sum-map(MapCursor): 'canada'                     22,080,899.3                    ±0.47%         177       59            5.50
✔   averageTwoValues(JsonCursor): 'canada'            7,711,584.2                    ±0.34%         574       82            5.43
✔   averageTwoValues(TextCursor): 'canada'           11,690,887.6                    ±0.51%         365       73            5.49
✔   averageTwoValues(MapCursor): 'canada'             9,904,137.9                    ±0.83%         444       74            5.49
✔   Direct: 'twitter'                                 6,097,720.3                    ±1.04%         747       83            5.42
✔   cursorToJsonObject(JsonCursor): 'twitter'        25,316,256.7                    ±2.74%         156       52            5.57
✔   cursorToJsonObject(TextCursor): 'twitter'        31,688,712.9                    ±1.15%         112       56            5.63
✔   cursorToJsonObject(MapCursor): 'twitter'         24,380,555.5                    ±1.28%         162       54            5.55
✔   jsonableTreeFromCursor(JsonCursor): 'twitter'    43,340,035.8                    ±6.71%          86       43            5.78
✔   jsonableTreeFromCursor(TextCursor): 'twitter'    46,183,078.0                    ±2.97%          78       39            5.77
✔   jsonableTreeFromCursor(MapCursor): 'twitter'     43,641,635.5                    ±5.33%          84       42            5.75
✔   mapTreeFromCursor(JsonCursor): 'twitter'         45,557,930.7                    ±5.80%          76       38            5.83
✔   mapTreeFromCursor(TextCursor): 'twitter'         53,828,455.8                    ±7.52%          70       35            5.96
✔   mapTreeFromCursor(MapCursor): 'twitter'          51,043,841.4                    ±9.29%          78       39            5.83
✔   sum(JsonCursor): 'twitter'                        8,948,613.3                    ±0.23%         504       72            5.43
✔   sum(TextCursor): 'twitter'                       14,597,055.6                    ±0.79%         284       71            5.49
✔   sum(MapCursor): 'twitter'                         9,374,745.1                    ±0.85%         468       78            5.42
✔   sum-map(JsonCursor): 'twitter'                   15,717,769.6                    ±0.25%         264       66            5.47
✔   sum-map(TextCursor): 'twitter'                   21,742,179.6                    ±0.55%         180       60            5.50
✔   sum-map(MapCursor): 'twitter'                    16,434,961.4                    ±0.68%         252       63            5.46
✔   averageTwoValues(JsonCursor): 'twitter'             132,943.7                    ±0.15%      38,003       89            5.38
✔   averageTwoValues(TextCursor): 'twitter'             190,555.7                    ±0.15%      26,522       89            5.41
✔   averageTwoValues(MapCursor): 'twitter'              134,036.6                    ±0.31%      36,363       93            5.45
```

After:

```
✔   Direct: 'canada'                                  5,370,747.6                    ±0.97%         891       81            5.33
✔   cursorToJsonObject(JsonCursor): 'canada'         26,005,786.5                    ±5.34%         156       52            5.73
✔   cursorToJsonObject(TextCursor): 'canada'         31,461,568.2                    ±4.45%         112       56            5.56
✔   cursorToJsonObject(MapCursor): 'canada'          32,018,202.5                    ±3.53%         110       55            5.62
✔   jsonableTreeFromCursor(JsonCursor): 'canada'     39,568,948.6                    ±6.15%          92       46            5.72
✔   jsonableTreeFromCursor(TextCursor): 'canada'     48,780,693.2                    ±4.79%          74       37            5.76
✔   jsonableTreeFromCursor(MapCursor): 'canada'      51,817,209.9                    ±6.18%          72       36            5.82
✔   mapTreeFromCursor(JsonCursor): 'canada'          57,198,626.8                    ±8.98%          66       33            5.99
✔   mapTreeFromCursor(TextCursor): 'canada'          66,225,592.8                    ±9.92%          43       43            5.70
✔   mapTreeFromCursor(MapCursor): 'canada'           61,833,861.4                    ±4.39%          42       42            5.66
✔   sum(JsonCursor): 'canada'                        10,387,735.5                    ±0.37%         410       82            5.42
✔   sum(TextCursor): 'canada'                        15,928,974.0                    ±0.42%         260       65            5.46
✔   sum(MapCursor): 'canada'                         16,586,827.3                    ±0.49%         252       63            5.56
✔   sum-map(JsonCursor): 'canada'                    18,627,366.1                    ±0.30%         207       69            5.44
✔   sum-map(TextCursor): 'canada'                    25,981,752.4                    ±0.33%         132       66            5.46
✔   sum-map(MapCursor): 'canada'                     23,939,566.7                    ±0.41%         165       55            5.54
✔   averageTwoValues(JsonCursor): 'canada'            7,398,649.2                    ±0.21%         602       86            5.46
✔   averageTwoValues(TextCursor): 'canada'           11,079,639.8                    ±0.71%         385       77            5.49
✔   averageTwoValues(MapCursor): 'canada'            10,206,601.5                    ±0.98%         444       74            5.52
✔   Direct: 'twitter'                                 6,018,562.1                    ±0.43%         756       84            5.41
✔   cursorToJsonObject(JsonCursor): 'twitter'        28,686,012.1                    ±1.60%         122       61            5.54
✔   cursorToJsonObject(TextCursor): 'twitter'        36,124,818.1                    ±1.89%          98       49            5.59
✔   cursorToJsonObject(MapCursor): 'twitter'         25,945,477.3                    ±1.25%         153       51            5.52
✔   jsonableTreeFromCursor(JsonCursor): 'twitter'    42,356,132.5                    ±5.18%          84       42            5.69
✔   jsonableTreeFromCursor(TextCursor): 'twitter'    50,940,568.5                    ±2.89%          72       36            5.94
✔   jsonableTreeFromCursor(MapCursor): 'twitter'     43,616,228.6                    ±4.03%          82       41            5.74
✔   mapTreeFromCursor(JsonCursor): 'twitter'         49,788,395.0                    ±6.59%          76       38            5.89
✔   mapTreeFromCursor(TextCursor): 'twitter'         55,116,249.6                    ±6.39%          68       34            5.70
✔   mapTreeFromCursor(MapCursor): 'twitter'          47,718,170.0                    ±7.42%          78       39            5.78
✔   sum(JsonCursor): 'twitter'                       13,582,252.5                    ±1.61%         350       70            5.44
✔   sum(TextCursor): 'twitter'                       19,404,282.9                    ±0.64%         201       67            5.50
✔   sum(MapCursor): 'twitter'                        11,362,900.2                    ±0.81%         375       75            5.42
✔   sum-map(JsonCursor): 'twitter'                   19,975,314.0                    ±0.23%         195       65            5.49
✔   sum-map(TextCursor): 'twitter'                   25,340,131.6                    ±0.65%         156       52            5.48
✔   sum-map(MapCursor): 'twitter'                    17,683,949.9                    ±0.56%         219       73            5.47
✔   averageTwoValues(JsonCursor): 'twitter'             122,824.1                    ±0.17%      40,945       95            5.39
✔   averageTwoValues(TextCursor): 'twitter'             177,810.2                    ±0.33%      28,179       93            5.43
✔   averageTwoValues(MapCursor): 'twitter'              142,460.7                    ±0.18%      35,399       91            5.40
```

Comment (CraigMacomber) [reply packages/dds/tree/src/feature-libraries/treeCursorUtils.ts:183]: I think it fine if you want to merge this as is, and do the perf tweaking in a followup PR.

Comment (jenn-le) [reply packages/dds/tree/src/feature-libraries/treeCursorUtils.ts]: Just pushed the change. It's significantly faster but I think you should review the change. I also made a change to a test case so that the round trip tests pass.

File: packages/dds/tree/src/feature-libraries/mapTreeCursor.ts [hunk 11]
```diff
@@ -24,7 +24,18 @@ export function singleMapTreeCursor(root: MapTree): ITreeCursorSynchronous {
 const adapter: CursorAdapter<MapTree> = {
     value: (node) => node.value,
     type: (node) => node.type,
-    keysFromNode: (node) => [...node.fields.keys()], // TODO: don't convert this to array here.
+    keysFromNode: (node) => {
```
Comment (CraigMacomber) [inline packages/dds/tree/src/feature-libraries/mapTreeCursor.ts:27]: like jsonableTree, all current users of MapTree should avoid storing empty field arrays, so we shouldn't need to do anything special here. I checked and that isn't documented on MapTree: you should be able to ad docs for that instead of handling it here.

Also, if you were going to keep this code, there is no reason for it to call isArray: the value type in MapTree is MapTree[]: you just found a slower way to cast away the fact that get can return undefined. A better way to do that would be to iterate over the map instead of its keys and calling get for each one.

**Review by CraigMacomber — APPROVED** [review]
